### PR TITLE
feat(draw_buf): support indexed image in draw_buf_copy

### PR DESCRIPTION
**Pull request by XuNeo**
### Description of the feature or fix

A clear and concise description of what the bug or new feature is.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.10](https://github.com/szepeviktor/astyle/releases/tag/v3.4.10) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

File: src/draw/lv_draw_buf.c [hunk 1]
```diff
@@ -121,9 +121,20 @@ void lv_draw_buf_copy(lv_draw_buf_t * dest, const lv_area_t * dest_area,
...
+    LV_ASSERT_FORMAT_MSG(dest->header.cf == src->header.cf, "Color format mismatch: %d != %d",
+                         dest->header.cf, src->header.cf);
+
     if(dest_area == NULL) line_width = dest->header.w;
     else line_width = lv_area_get_width(dest_area);
 
+    /* For indexed image, copy the palette if we are copying full image area*/
+    if(dest_area == NULL || src_area == NULL) {
+        if(LV_COLOR_FORMAT_IS_INDEXED(dest->header.cf)) {
+            lv_memcpy(dest->data, src->data, LV_COLOR_INDEXED_PALETTE_SIZE(dest->header.cf) * 4);
```
Comment (FASTSHIFT) [inline src/draw/lv_draw_buf.c]: 4 -> sizeof(lv_color32_t)

**Review by FASTSHIFT — CHANGES_REQUESTED** [review]

File: src/draw/lv_draw_buf.c [hunk 2]
```diff
@@ -279,10 +290,13 @@ void * lv_draw_buf_goto_xy(const lv_draw_buf_t * buf, uint32_t x, uint32_t y)
     LV_ASSERT_NULL(buf);
     if(buf == NULL) return NULL;
 
-    uint8_t * data = buf->data + buf->header.stride * y;
+    uint8_t * data = buf->data;
+
+    /*Skip palette*/
+    data += LV_COLOR_INDEXED_PALETTE_SIZE(buf->header.cf) * 4;
```
Comment (FASTSHIFT) [inline src/draw/lv_draw_buf.c]: 4 -> sizeof(lv_color32_t))

**Review by FASTSHIFT — CHANGES_REQUESTED** [review]

**Review by FASTSHIFT — APPROVED** [review]

**Review by kisvegabor — COMMENTED** [review]
Looks good, but could you add a test for it?

**Comment by XuNeo** [comment]
Hi Gabor,

The test can be covered in https://github.com/lvgl/lvgl/pull/5690/files#diff-4cda730c9bc038547013a81b7cf252aaafe3cd9bb10e5230e6d181fbc6bd71daR89 which will check buffer copy for all supported color formats.

**Comment by kisvegabor** [comment]
Awesome!

**Review by kisvegabor — APPROVED** [review]

**Pull request merged**